api/apikeys.go

```diff
@@ -3,16 +3,91 @@ package api
 import (
 	"net/http"
 	"strconv"
+	"time"
 
 	"github.com/SkynetLabs/skynet-accounts/database"
 	"github.com/julienschmidt/httprouter"
 	"gitlab.com/NebulousLabs/errors"
+	"go.mongodb.org/mongo-driver/bson/primitive"
 	"go.mongodb.org/mongo-driver/mongo"
 )
 
+type (
+	// APIKeyPOST describes the body of a POST request that creates an API key
+	APIKeyPOST struct {
+		Public   bool     `json:"public,string"`
+		Skylinks []string `json:"skylinks"`
+	}
+	// APIKeyPUT describes the request body for updating an API key
+	APIKeyPUT struct {
+		Skylinks []string
+	}
+	// APIKeyPATCH describes the request body for updating an API key by
+	// providing only the requested changes
+	APIKeyPATCH struct {
+		Add    []string
+		Remove []string
+	}
+	// APIKeyResponse is an API DTO which mirrors database.APIKey.
+	APIKeyResponse struct {
+		ID        primitive.ObjectID `json:"id"`
+		UserID    primitive.ObjectID `json:"-"`
+		Public    bool               `json:"public,string"`
+		Key       database.APIKey    `json:"-"`
+		Skylinks  []string           `json:"skylinks"`
+		CreatedAt time.Time          `json:"createdAt"`
+	}
+	// APIKeyResponseWithKey is an API DTO which mirrors database.APIKey but
+	// also reveals the value of the Key field. This should only be used on key
+	// creation.
+	APIKeyResponseWithKey struct {
+		APIKeyResponse
+		Key database.APIKey `json:"key"`
+	}
+)
+
+// Valid checks if the request and its parts are valid.
+func (akp APIKeyPOST) Valid() bool {
```

Comment:
so this is nice but I think we might benefit from an error providing some context here,
because of the fact that `akp.Skylinks` can be huge and it will be annoying for the
caller to find out which one is giving him issues